demo_apps/whatsapp_llama2.md

```diff
@@ -0,0 +1,151 @@
+
+```
+import langchain
+from langchain.llms import Replicate
+from flask import Flask
+from flask import request
+import os
+import requests
+import json
+
+class WhatsAppClient:
+    API_URL = "https://graph.facebook.com/v17.0/"
+    WHATSAPP_API_TOKEN = "<Temporary access token from your WhatsApp API Setup>"
+    WHATSAPP_CLOUD_NUMBER_ID = "<Phone number ID from your WhatsApp API Setup>"
+
+    def __init__(self):
+        self.headers = {
+            "Authorization": f"Bearer {self.WHATSAPP_API_TOKEN}",
+            "Content-Type": "application/json",
+        }
+        self.API_URL = self.API_URL + self.WHATSAPP_CLOUD_NUMBER_ID
+        
+    def send_text_message(self, message, phone_number):
+        payload = {
+            "messaging_product": 'whatsapp',
+            "to": phone_number,
+            "type": "text",
+            "text": {
+                "preview_url": False,
+                "body": message
+            }
+        }
+        response = requests.post(f"{self.API_URL}/messages", json=payload, headers=self.headers)
+        return response.status_code
+```
+
+Finally, add the code below to llama-chatbot.py, which creates a Llama 2 instance and defines an HTTP method `msgrcvd` to:
+1. receive the user message forwarded by the webhook;
+2. ask Llama 2 for the answer;
+3. call the `WhatsAppClient`'s `send_text_message`` with a recipient's phone number.
+   
+```   
+os.environ["REPLICATE_API_TOKEN"] = "<your replicate api token>"    
+llama2_13b_chat = "meta/llama-2-13b-chat:f4e2de70d66816a838a89eeeb621910adffb0dd0baba3976c96980970978018d"
+
+llm = Replicate(
+    model=llama2_13b_chat,
+    model_kwargs={"temperature": 0.01, "top_p": 1, "max_new_tokens":500}
+)
+client = WhatsAppClient()
+app = Flask(__name__)
+
+@app.route("/")
+def hello_llama():
+    return "<p>Hello Llama 2</p>"
+
+@app.route('/msgrcvd', methods=['POST', 'GET'])
+def msgrcvd():    
+    message = request.args.get('message')
+    answer = llm(message)
+    client.send_text_message(answer, "<a recipient phone number from your WhatsApp API Setup>")
+    return message + "<p/>" + answer
+```
+
+Now it's time to modify the webhook to complete the whole app.
+
+## Modifying the Webhook 
+
+Open your glitch.com webhook URL created earlier, and after the code snippet:
+
+```
+// message received! 
+console.log(req.body["entry"][0]["changes"][0]["value"]["messages"][0]["text"]["body"]);
+```
+
+add the code below - remember to change <web server public IP>, which needs to be publicly visible, to the IP of the server where your Llama 2 enabled web app in the previous section runs:
+
+```
+  let url = "http://<web server public IP>:5000/msgrcvd?message=" + 
+    req.body["entry"][0]["changes"][0]["value"]["messages"][0]["text"]["body"]
+
+  axios.get(url)
+    .then(response => {
+      console.log(response.data);
```

Comment:
From this code, it's not clear that the response gets sent back to the user via WA. It might be good to add the full code here, for reference.

Reply:
Good point - but actually the following code the app.js should be deleted because "the response gets sent back to the user via WA" is now done in the llama_chatbot.py. I updated the tutorial on this.